plugins/urls/__init__.py

```diff
@@ -335,343 +355,151 @@ def _respond_shorten(self, result, source, handler):
-                return self.parse_title(headers["location"])
-
-            ct = headers_dict["content-type"]
-            if ";" in ct:
-                ct = ct.split(";")[0]
-
-            self.logger.trace(_("Content-type: %s") % repr(ct))
-
-            if ct not in self.content_types:
-                self.logger.debug(_("Content-type is not allowed."))
-                return None, None
-
-            page = response.read()
-            soup = BeautifulSoup(page)
-            if soup.title and soup.title.string:
-                title = soup.title.string.strip()
-                title = re.sub("\s+", " ", title)
-                title = to_unicode(title)
-                domain = to_unicode(domain)
-                return title, domain
-            else:
-                return None, None
-        except Exception as e:
-            if not str(e).lower() == "not viewing html":
-                self.logger.exception(_("Error parsing title."))
-                return str(e), domain
-            return None, None
-
-    def _shorten(self, txn, url, handler):
-        """
-        Shorten a URL, checking the database in case it's already been
-        done. This is a database interaction and uses Deferreds.
-        """
-
-        txn.execute("SELECT * FROM urls WHERE url=? AND shortener=?",
-                    (url, handler.lower()))
-        r = txn.fetchone()
-
-        self.logger.trace(_("Result (SQL): %s") % repr(r))
-
-        if r is not None:
-            return r[2]
-
-        if handler in self.shorteners:
-            result = self.shorteners[handler](url)
-
-            txn.execute("INSERT INTO urls VALUES (?, ?, ?)",
-                        (url, handler.lower(), result))
-            return result
-        return None
-
-    def shorten_url(self, url, handler):
-        """
-        Shorten a URL using the specified handler. This returns a Deferred.
-
-        :param url: The URL to shorten
-        :param handler: The name of the handler to shorten with
-
-        :type url: str
-        :type handler: str
-
-        :returns: Deferred which will fire with the result or None
-        :rtype: Deferred
-        """
+                        d = handler.call(_url, context)
+                    except Exception as e:
+                        self.logger.exception(
+                            "Error caught while attempting to call "
+                            "handler {0}".format(
+                                handler.name
+                            )
+                        )
+
+                        continue
+
+                    if isinstance(d, Deferred):
+                        r = yield d
+                    else:
+                        r = d
+                    if not r:
+                        return
 
-        self.logger.trace(_("URL: %s") % url)
-        self.logger.trace(_("Handler: %s") % handler)
+    def add_handler(self, handler, priority=0):
```

Comment:
What do we have for the event manager? Actually, *do* we have any for the event manager?